Provide the (reconstructed) input from the parser

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Provide the input source from the parser

nilsdeppe commented on src/Options/ParseOptions.hpp:
```diff
@@ -36,6 +36,14 @@
 #include "Utilities/TypeTraits/IsStdArrayOfSize.hpp"
 
 namespace Options {
+/// Option parser tag to retrieve the YAML source.  This tag can be
```
Question: would it be reasonable to preserve the source to keep the comments? Or maybe yaml-cpp supports that (less likely)? I'm just thinking of the case where someone might actually care about the comments in the input file. If that's really difficult to do, then we can just ignore it :)

nilsvu commented:
@wthrowe Would you mind rebasing on develop so the one renamed check runs correctly?

wthrowe commented:
Rewrote to read the file into a string so we can return it exactly.  Squashed directly because it was a near complete rewrite (except for the test, which is unchanged).  Also rebased on develop.

nilsdeppe reviewed (dismissed):
This is awesome, thank you so much :D

nilsvu commented on src/Options/ParseOptions.hpp:
```diff
@@ -431,6 +446,14 @@ struct get_impl<Tag, Metavariables, Tag> {
+template <typename Metavariables>
+struct get_impl<InputSource, Metavariables, InputSource> {
+  template <typename OptionList, typename Group>
+  static InputSource::type apply(const Parser<OptionList, Group>& opts) {
```
Should this return by const-ref?

wthrowe replied:
The function wrapping this forwards it on by value, so no point.

nilsvu commented on src/Options/ParseOptions.hpp:
```diff
@@ -273,22 +283,27 @@ template <typename OptionList, typename Group>
-    ERROR("Could not open the input file " << file_name);
-  } catch (const YAML::Exception& e) {
+    parse(YAML::Load(input_source_));
+  } catch (YAML::Exception& e) {
```
Did you remove the `const` here for a particular reason? AFAICT the `parser_error` function takes its argument by const-ref

wthrowe replied:
Because I copied it from the next function where it was already missing.  I've fixed them both.  (Only change in the force push.)
